src/ring_buffer.zig

```diff
@@ -90,6 +99,13 @@ pub fn RingBuffer(comptime T: type, comptime size: usize) type {
             return result;
         }
 
+        /// Remove and return the last item, if any.
+        pub fn pop_tail(self: *Self) ?T {
+            const result = self.tail() orelse return null;
+            self.count -= 1;
```

Comment:
- Maybe instead of `advance_head` and `retreat_tail` we could have `drop_head` and `drop_tail` (or s/drop/discard).
- Tangentially related: `advance_tail` is a public function — but isn't it unsafe to use (except as `push` does internally)? i.e. outside of `push`, calling it would allow access to old memory.

Reply:
> Tangentially related: advance_tail is a public function — but isn't it unsafe to use (except as push does internally)? i.e. outside of push, calling it would allow access to old memory.

The high-level RingBuffer methods like `push` use `advance_tail()` internally, but we also use it directly together with `next_tail_ptr()` like so:

```zig
if (it.tables.next_tail_ptr()) |next_tail| {
    read_next_table(next_tail);
    it.tables.advance_tail();
    return true;
} else {
    const table = it.tables.head_ptr().?;
    while (table.peek() != null) {
        it.values.push(table.pop()) catch unreachable;
    }
    it.tables.advance_head();

    read_next_table(it.tables.next_tail_ptr().?);
    it.tables.advance_tail();
    return true;
}
```

Reply:
> Maybe instead of advance_head and retreat_tail we could have drop_head and drop_tail (or s/drop/discard).

I like `advance` because it has the idea of movement, moving the head or tail pointer. I got it from liburing, e.g. `io_uring_cq_advance(&ring, 4);`. `retreat` is the antonym also with the same number of characters so they're symmetrical and line up in the source.